torch/_dynamo/backends/train_step.py

```diff
@@ -0,0 +1,111 @@
+import warnings
+from typing import List
+
+from functorch import functionalize
+
+import torch
+import torch.fx.traceback as fx_traceback
+import torch.utils._pytree as pytree
+from torch import fx
+from torch._dynamo import register_backend
+from torch._dynamo.backends.registry import lookup_backend
+from torch._subclasses.fake_tensor import FakeTensor, FakeTensorMode
+from torch.fx.experimental.proxy_tensor import make_fx
+from torch.fx.interpreter import Interpreter
+from torch.nn.utils import stateless
+
+
+def train_step_compiler(backend_compile_fn):
+    """Note [Train Step Compile]
+
+    Usually, torch.compile() allows graph-breaks and compiles pairs of forward (+backward) by
+    extracting sections of forward from python programs and using AotAutograd to produce corresponding
+    chunks of backwards, tying it back together with an AotFunction.
+
+    Instead, TrainStepCompiler assumes the user compiles a full train_step function complete with calls to
+    .backward(), optimizer step(), and zero_grad().  It additionally requires no graph-breaks.
+
+    Args:
+        backend_compile_fn (callable): A dynamo compiler function, to be invoked to compile each subgraph.
+    """
+
+    def _compile_fn(mod: fx.GraphModule, fake_inputs: List[torch.Tensor]):
+        """
+        Step 1: Assert inputs (from user) are already Fake, and user their FakeTensorMode
+                (created by dynamo) to fakeify the module's parameters
+        """
+        torch.set_grad_enabled(True)
+        torch._dynamo.utils.assert_no_fake_params_or_buffers(mod)
+        assert len(fake_inputs) > 0, "Expected at least one input"
+        fake_mode = fake_inputs[0].fake_mode
+        assert isinstance(
+            fake_mode, FakeTensorMode
+        ), "Expected a valid FakeTensorMode on dynamo inputs"
+
+        def fakeify_inputs(flat_args):
+            def convert(idx, x):
+                # todo: do we expect symint inputs?
+                assert isinstance(x, torch.Tensor)
+                return fake_mode.from_tensor(x, static_shapes=False)
+
+            return [convert(idx, x) for idx, x in enumerate(flat_args)]
+
+        params = {
+            **dict(mod.named_parameters(remove_duplicate=False)),
+            **dict(mod.named_buffers(remove_duplicate=False)),
+        }
+        params_flat, params_spec = pytree.tree_flatten(params)
+        params_len = len(params_flat)
+        fake_params_flat = fakeify_inputs(params_flat)
+        full_fake_args = fake_params_flat + fake_inputs
+
+        """
+        Step 2: Create a new graphmodule that accepts parameters and user-inputs
+                as inputs to forward(), instead of accessing parameters as attrs.
+        """
+
+        def functional_call(*args, **kwargs):
+            with stateless._reparametrize_module(
```

Comment:
I assume we need to reparameterize optimizer states as well, as soon as we use stateful ones?

Like this: 
https://github.com/pytorch/pytorch/blob/482f87a7bc205d8a51744b382002eb23d4e401f2/torch/distributed/_spmd/api.py#L474

Reply:
thanks, still fighting with dynamo to get the optimizer into the graph.  But will try to use your hint when i get there.